### DISCUSSION TITLE
Studying radio frequency and SNR of devices observed by rtl_433

### DISCUSSION THREAD

**Discussion by hdtodd** [merbanan/rtl_433#2438] [Show and tell]
I've been analyzing rtl_433 JSON packet information on a device-by-device basis to try to understand better how to characterize the devices seen by the RTL-SDR dongle and decoded by rtl_433.  In case it's useful for anyone else, I'm sharing what I observed here.

Using data from a colleague who lives in a high-traffic area, I extracted all JSON records from a 2-day log file into separate files, converted to CSV, and imported it into Excel so I could study the packet patterns.  In the course of doing that, I thought I'd try try to identify visually any characteristics that I could try to program to recognize in my transmission-analysis program ([rtl_snr](https://github.com/hdtodd/rtl_snr) ... check out the v2.0.0 branch  and v2-README if you're curious). 

I wanted to see if there were any obvious indications of radio frequency variation over time or any indications of SNR-Freq variation correlation.  So I did two plots for each:

- Freq over time line plot
- SNR vs Freq x-y scatter.

I used my colleague's data for an Acurite-986, id 46015, which is in a freezer, so the signal is relatively attenuated, and for a LaCrosse-TX, id 61, which is located in his back yard.  The graphs are below.

I suspect the differences have more to do with device manufacturers and age than anything else, but I did see three curiosities:

- From the first pair of graphs, the substantial excursions beyond baseline were generally lower frequencies for the Acurite and higher frequencies for the LaCrosse.
- From the second pair of graphs, the Acurite (in the freezer) were not as tightly clustered around the baseline frequency as the LaCrosse (backyard) frequencies were.
- From the second pair of graphs, variation in frequency from baseline correlated with higher SNR for the Acurite (freezer) whereas the SNRs were roughly the same across the range of frequencies for the LaCrosse (backyard).

If you have any other insights that might help interpret the similarities and differences in devices and their environments based upon the data as seen by a remote observer, please post!!!

David

<img width="914" alt="image" src="https://user-images.githubusercontent.com/5464284/226460468-54d7e1ab-d939-4db4-8eaf-f008660c401e.png">

<img width="849" alt="image" src="https://user-images.githubusercontent.com/5464284/226460650-94813738-d224-4fed-9ff0-74481e077677.png">

<img width="914" alt="image" src="https://user-images.githubusercontent.com/5464284/226460716-0deda720-b564-4475-b0fa-a2882751c159.png">

<img width="857" alt="image" src="https://user-images.githubusercontent.com/5464284/226460793-ca9986d8-6021-4ea5-a198-be89eca8b5e2.png">

**Comment by merbanan** [comment]
Both receiver tuner and transmitting sensor frequency drift by temperature. So you need to keep one stable to get something relevant out of the data.

**Reply by rct** [reply]
> Both receiver tuner and transmitting sensor frequency drift by temperature. So you need to keep one stable to get something relevant out of the data.

Actually I was expecting to see somewhat wider frequency variance.   Here's the data from David's SNR.py

```
Processed 79604 Packets as 35000 De-Duplicated Transmissions in  0.71sec
Packets dated from Mon 2023-03-06 12:22:50 to Wed 2023-03-08 08:01:57

                                         Signal-to-Noise                        Frequency (MHz)
                               _________________________________    ____________________________________________
Device                          #Pkts   Mean ±   𝜎    Min    Max     #Pkts     Mean  ±   𝜎          Min      Max

Acurite-986/46015/2F             2341    9.1 ±  1.9   5.5   26.1      2341   433.947 ±  0.009   433.793  434.006
LaCrosse-TX/61/                 10075   21.7 ±  1.5  17.4   26.9     10075   433.826 ±  0.005   433.793  434.004
```

The Acurite sensor is in a refrigerator so it sees small but very regular temperature cycles of a few degrees.

![image](https://user-images.githubusercontent.com/38770/226472671-c8a9cd0d-8334-43a0-9ed6-3e19e30a562a.png)

The LaCrosse sensor is outside. During this period it saw around a 25F range of temperatures.

![image](https://user-images.githubusercontent.com/38770/226473046-65ccf63f-79a2-4443-9b75-6809c38cff80.png)

The RTL-SDR is in a relatively temperature controlled environment and is running full time. Though a fair point is the RTL-SDR that recorded that data isn't one of the newer ones with TCXO. 

The things that are curious to me are the outliers.  Particularly as David has pointed out, the Acurite has a number of successful decodes at might higher SNRs with a great frequency range.  (The signal from the Acurite is somewhat attenuated by being inside of a refrigerator.) 

One idea is maybe that could possibly be some artifact of the signal measurement/estimation code, possibly the influence of other strong signals that might be part of the same buffer?

**Reply by zuckschwerdt** [reply]
The frequency measurement is not particularly accurate. I hope to add a more precise FFT mode soon to improve that (if you have the cpu load to spend).
